sdk/trace: use sync.Pool in randomIDGenerator instead of Mutex

maisem opened the pull request:
randomIDGenerator used a sync.Mutex to coordinate generation of TraceID
and SpanID. When used in relatively hot code paths this would cause
significant mutex contention as observed in pprof/mutex profiles.

This uses a sync.Pool instead which allows using generating and using
of *rand.Rand as needed without the need for a Lock.

```
name                 old time/op    new time/op    delta
SpanIDGeneration-10     191ns ± 5%      64ns ± 2%  -66.60%  (p=0.000 n=9+10)

name                 old alloc/op   new alloc/op   delta
SpanIDGeneration-10     0.00B          0.00B          ~     (all equal)

name                 old allocs/op  new allocs/op  delta
SpanIDGeneration-10      0.00           0.00          ~     (all equal)
```

Signed-off-by: Maisem Ali <maisem@tailscale.com>

dmathieu commented:
Could you add a benchmark for this? It would then be nice to run it on the main branch, and with your changes and paste the results within this PR.
You can find examples of other benchmarks in the same package here: https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/benchmark_test.go

maisem commented:
@dmathieu I've added a benchmark. PTAL

maisem commented:
Minor update to fix the lint check.

MadVikingGod reviewed (changes requested):
When benchmarking this change with the current suite of benchmarks, there is some major performance regression.  I think those need to be addressed or explained before this would be acceptable.

```
$ benchstat bench-main bench-3134 
name                                          old time/op    new time/op    delta
SpanProcessor-4                                 13.1µs ± 4%    16.6µs ±35%  +26.70%  (p=0.003 n=10+10)
SpanProcessorVerboseLogging-4                   13.7µs ± 6%    17.1µs ±31%  +25.45%  (p=0.002 n=10+10)
SpanLimits/AttributeValueLengthLimit-4          10.7µs ± 6%    13.6µs ±34%  +27.31%  (p=0.017 n=9+10)
SpanLimits/AttributeCountLimit-4                9.12µs ± 3%   12.20µs ±40%  +33.66%  (p=0.000 n=9+10)
SpanLimits/EventCountLimit-4                    10.2µs ± 4%    12.9µs ±34%  +25.77%  (p=0.015 n=10+10)
SpanLimits/LinkCountLimit-4                     10.1µs ± 5%    13.0µs ±39%  +28.87%  (p=0.011 n=10+10)
SpanLimits/AttributePerEventCountLimit-4        10.7µs ± 8%    14.0µs ±43%     ~     (p=0.052 n=10+10)
SpanLimits/AttributePerLinkCountLimit-4         10.5µs ± 7%    13.3µs ±36%  +25.86%  (p=0.009 n=10+10)
SpanSetAttributesOverCapacity-4                 2.24µs ± 4%    2.55µs ±22%  +13.93%  (p=0.001 n=9+10)
StartEndSpan/AlwaysSample-4                      690ns ± 7%     818ns ±41%  +18.55%  (p=0.035 n=10+10)
StartEndSpan/NeverSample-4                       321ns ± 7%     396ns ±34%  +23.45%  (p=0.000 n=9+10)
SpanWithAttributes_4/AlwaysSample-4             1.37µs ± 2%    1.70µs ±36%  +24.60%  (p=0.004 n=9+10)
SpanWithAttributes_4/NeverSample-4               501ns ± 6%     604ns ±35%  +20.71%  (p=0.002 n=10+10)
SpanWithAttributes_8/AlwaysSample-4             1.94µs ± 6%    2.41µs ±41%  +24.15%  (p=0.007 n=10+10)
SpanWithAttributes_8/NeverSample-4               662ns ± 2%     831ns ±44%  +25.62%  (p=0.000 n=9+10)
SpanWithAttributes_all/AlwaysSample-4           1.72µs ± 4%    2.14µs ±51%  +24.56%  (p=0.000 n=10+10)
SpanWithAttributes_all/NeverSample-4             546ns ± 3%     666ns ±37%  +22.03%  (p=0.001 n=8+10)
SpanWithAttributes_all_2x/AlwaysSample-4        2.63µs ± 6%    3.32µs ±48%  +26.31%  (p=0.004 n=9+10)
SpanWithAttributes_all_2x/NeverSample-4          755ns ± 2%    1022ns ±45%  +35.38%  (p=0.000 n=9+10)
SpanWithEvents_4/AlwaysSample-4                 1.66µs ± 5%    2.09µs ±28%  +25.40%  (p=0.001 n=10+10)
SpanWithEvents_4/NeverSample-4                   333ns ± 7%     412ns ±27%  +23.77%  (p=0.001 n=10+10)
SpanWithEvents_8/AlwaysSample-4                 2.53µs ± 3%    3.24µs ±41%  +28.22%  (p=0.001 n=8+10)
SpanWithEvents_8/NeverSample-4                   333ns ± 3%     422ns ±29%  +26.76%  (p=0.000 n=9+10)
SpanWithEvents_WithStackTrace/AlwaysSample-4    1.02µs ± 6%    1.25µs ±28%  +21.65%  (p=0.008 n=10+10)
SpanWithEvents_WithStackTrace/NeverSample-4      363ns ± 4%     470ns ±35%  +29.39%  (p=0.000 n=9+10)
SpanWithEvents_WithTimestamp/AlwaysSample-4     1.01µs ± 3%    1.25µs ±30%  +24.00%  (p=0.001 n=9+10)
SpanWithEvents_WithTimestamp/NeverSample-4       424ns ± 9%     523ns ±28%     ~     (p=0.063 n=10+10)
TraceID_DotString-4                             86.6ns ± 2%   105.6ns ±43%  +21.86%  (p=0.000 n=9+10)
SpanID_DotString-4                              62.7ns ± 3%    73.3ns ±31%  +16.96%  (p=0.001 n=9+10)
```

maisem commented:
@MadVikingGod I can't repro that. It almost looks like you had some CPU contention during the second run, can you run it again?

I see a difference but its within reason
```
name                                           old time/op    new time/op    delta
SpanProcessor-10                                 8.23µs ± 8%    8.53µs ± 8%    ~     (p=0.165 n=10+10)
SpanProcessorVerboseLogging-10                   8.44µs ± 4%    8.88µs ± 7%  +5.19%  (p=0.001 n=10+10)
SpanLimits/AttributeValueLengthLimit-10          5.58µs ± 2%    5.57µs ± 0%    ~     (p=1.000 n=10+8)
SpanLimits/AttributeCountLimit-10                4.88µs ± 3%    4.85µs ± 1%    ~     (p=0.184 n=10+10)
SpanLimits/EventCountLimit-10                    5.22µs ± 2%    5.38µs ± 3%  +3.00%  (p=0.007 n=10+10)
SpanLimits/LinkCountLimit-10                     5.18µs ± 2%    5.42µs ± 5%  +4.57%  (p=0.000 n=10+10)
SpanLimits/AttributePerEventCountLimit-10        5.37µs ± 2%    5.38µs ± 2%    ~     (p=0.931 n=9+9)
SpanLimits/AttributePerLinkCountLimit-10         5.35µs ± 3%    5.39µs ± 3%    ~     (p=0.234 n=10+9)
SpanSetAttributesOverCapacity-10                 2.36µs ± 1%    2.38µs ± 0%  +0.80%  (p=0.007 n=10+10)
StartEndSpan/AlwaysSample-10                      480ns ± 1%     498ns ± 1%  +3.80%  (p=0.000 n=8+9)
StartEndSpan/NeverSample-10                       233ns ± 1%     247ns ± 1%  +5.95%  (p=0.000 n=9+10)
SpanWithAttributes_4/AlwaysSample-10              846ns ± 2%     861ns ± 1%  +1.75%  (p=0.002 n=9+9)
SpanWithAttributes_4/NeverSample-10               334ns ± 2%     344ns ± 1%  +3.06%  (p=0.000 n=10+9)
SpanWithAttributes_8/AlwaysSample-10             1.15µs ± 3%    1.16µs ± 2%    ~     (p=0.089 n=10+10)
SpanWithAttributes_8/NeverSample-10               407ns ± 3%     425ns ± 4%  +4.56%  (p=0.000 n=10+10)
SpanWithAttributes_all/AlwaysSample-10           1.02µs ± 4%    1.02µs ± 2%    ~     (p=0.065 n=9+9)
SpanWithAttributes_all/NeverSample-10             349ns ± 1%     368ns ± 3%  +5.43%  (p=0.000 n=10+10)
SpanWithAttributes_all_2x/AlwaysSample-10        1.49µs ± 3%    1.48µs ± 2%    ~     (p=0.889 n=10+9)
SpanWithAttributes_all_2x/NeverSample-10          452ns ± 3%     465ns ± 1%  +2.74%  (p=0.000 n=10+8)
SpanWithEvents_4/AlwaysSample-10                 1.09µs ± 2%    1.11µs ± 2%  +2.38%  (p=0.014 n=9+10)
SpanWithEvents_4/NeverSample-10                   238ns ± 0%     254ns ± 0%  +6.73%  (p=0.000 n=8+8)
SpanWithEvents_8/AlwaysSample-10                 1.61µs ± 1%    1.64µs ± 1%  +1.44%  (p=0.000 n=10+9)
SpanWithEvents_8/NeverSample-10                   244ns ± 0%     260ns ± 1%  +6.38%  (p=0.000 n=9+10)
SpanWithEvents_WithStackTrace/AlwaysSample-10     655ns ± 1%     674ns ± 1%  +3.03%  (p=0.000 n=10+9)
SpanWithEvents_WithStackTrace/NeverSample-10      261ns ± 0%     279ns ± 2%  +6.95%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/AlwaysSample-10      655ns ± 1%     684ns ± 2%  +4.41%  (p=0.000 n=9+10)
SpanWithEvents_WithTimestamp/NeverSample-10       297ns ± 0%     314ns ± 0%  +5.93%  (p=0.000 n=10+8)
TraceID_DotString-10                             51.6ns ± 0%    50.4ns ± 2%  -2.26%  (p=0.000 n=10+9)
SpanID_DotString-10                              40.1ns ± 0%    40.2ns ± 0%  +0.28%  (p=0.001 n=9+9)

name                                           old alloc/op   new alloc/op   delta
SpanProcessor-10                                 8.75kB ± 0%    8.79kB ± 0%  +0.40%  (p=0.000 n=10+10)
SpanProcessorVerboseLogging-10                   9.90kB ± 0%    9.94kB ± 0%  +0.38%  (p=0.000 n=10+10)
SpanLimits/AttributeValueLengthLimit-10          10.6kB ± 0%    10.6kB ± 0%  +0.24%  (p=0.000 n=10+10)
SpanLimits/AttributeCountLimit-10                8.58kB ± 0%    8.60kB ± 0%  +0.25%  (p=0.000 n=10+10)
SpanLimits/EventCountLimit-10                    10.3kB ± 0%    10.3kB ± 0%  +0.26%  (p=0.000 n=10+10)
SpanLimits/LinkCountLimit-10                     10.3kB ± 0%    10.3kB ± 0%  +0.24%  (p=0.000 n=10+9)
SpanLimits/AttributePerEventCountLimit-10        10.5kB ± 0%    10.5kB ± 0%  +0.24%  (p=0.000 n=10+10)
SpanLimits/AttributePerLinkCountLimit-10         10.5kB ± 0%    10.5kB ± 0%  +0.23%  (p=0.000 n=10+10)
SpanSetAttributesOverCapacity-10                   496B ± 0%      497B ± 0%  +0.20%  (p=0.000 n=10+10)
StartEndSpan/AlwaysSample-10                       432B ± 0%      433B ± 0%  +0.23%  (p=0.000 n=10+10)
StartEndSpan/NeverSample-10                        128B ± 0%      128B ± 0%    ~     (all equal)
SpanWithAttributes_4/AlwaysSample-10             1.14kB ± 0%    1.14kB ± 0%  +0.18%  (p=0.000 n=10+10)
SpanWithAttributes_4/NeverSample-10                384B ± 0%      384B ± 0%    ~     (all equal)
SpanWithAttributes_8/AlwaysSample-10             1.90kB ± 0%    1.91kB ± 0%  +0.23%  (p=0.000 n=10+10)
SpanWithAttributes_8/NeverSample-10                640B ± 0%      641B ± 0%  +0.16%  (p=0.000 n=10+10)
SpanWithAttributes_all/AlwaysSample-10           1.71kB ± 0%    1.72kB ± 0%  +0.18%  (p=0.000 n=10+10)
SpanWithAttributes_all/NeverSample-10              448B ± 0%      449B ± 0%  +0.16%  (p=0.003 n=10+10)
SpanWithAttributes_all_2x/AlwaysSample-10        3.06kB ± 0%    3.06kB ± 0%  +0.23%  (p=0.000 n=10+8)
SpanWithAttributes_all_2x/NeverSample-10           768B ± 0%      770B ± 0%  +0.26%  (p=0.000 n=10+10)
SpanWithEvents_4/AlwaysSample-10                   864B ± 0%      866B ± 0%  +0.23%  (p=0.000 n=10+10)
SpanWithEvents_4/NeverSample-10                    128B ± 0%      128B ± 0%    ~     (all equal)
SpanWithEvents_8/AlwaysSample-10                 1.31kB ± 0%    1.31kB ± 0%  +0.23%  (p=0.000 n=10+10)
SpanWithEvents_8/NeverSample-10                    128B ± 0%      128B ± 0%    ~     (all equal)
SpanWithEvents_WithStackTrace/AlwaysSample-10      544B ± 0%      545B ± 0%  +0.18%  (p=0.000 n=10+10)
SpanWithEvents_WithStackTrace/NeverSample-10       144B ± 0%      144B ± 0%    ~     (all equal)
SpanWithEvents_WithTimestamp/AlwaysSample-10       568B ± 0%      569B ± 0%  +0.18%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/NeverSample-10        168B ± 0%      168B ± 0%    ~     (all equal)

name                                           old allocs/op  new allocs/op  delta
SpanProcessor-10                                   35.0 ± 0%      35.0 ± 0%    ~     (all equal)
SpanProcessorVerboseLogging-10                     41.0 ± 0%      41.0 ± 0%    ~     (all equal)
SpanLimits/AttributeValueLengthLimit-10            60.0 ± 0%      60.0 ± 0%    ~     (all equal)
SpanLimits/AttributeCountLimit-10                  54.0 ± 0%      54.0 ± 0%    ~     (all equal)
SpanLimits/EventCountLimit-10                      55.0 ± 0%      55.0 ± 0%    ~     (all equal)
SpanLimits/LinkCountLimit-10                       55.0 ± 0%      55.0 ± 0%    ~     (all equal)
SpanLimits/AttributePerEventCountLimit-10          58.0 ± 0%      58.0 ± 0%    ~     (all equal)
SpanLimits/AttributePerLinkCountLimit-10           58.0 ± 0%      58.0 ± 0%    ~     (all equal)
SpanSetAttributesOverCapacity-10                   3.00 ± 0%      3.00 ± 0%    ~     (all equal)
StartEndSpan/AlwaysSample-10                       2.00 ± 0%      2.00 ± 0%    ~     (all equal)
StartEndSpan/NeverSample-10                        2.00 ± 0%      2.00 ± 0%    ~     (all equal)
SpanWithAttributes_4/AlwaysSample-10               6.00 ± 0%      6.00 ± 0%    ~     (all equal)
SpanWithAttributes_4/NeverSample-10                3.00 ± 0%      3.00 ± 0%    ~     (all equal)
SpanWithAttributes_8/AlwaysSample-10               7.00 ± 0%      7.00 ± 0%    ~     (all equal)
SpanWithAttributes_8/NeverSample-10                3.00 ± 0%      3.00 ± 0%    ~     (all equal)
SpanWithAttributes_all/AlwaysSample-10             7.00 ± 0%      7.00 ± 0%    ~     (all equal)
SpanWithAttributes_all/NeverSample-10              3.00 ± 0%      3.00 ± 0%    ~     (all equal)
SpanWithAttributes_all_2x/AlwaysSample-10          8.00 ± 0%      8.00 ± 0%    ~     (all equal)
SpanWithAttributes_all_2x/NeverSample-10           3.00 ± 0%      3.00 ± 0%    ~     (all equal)
SpanWithEvents_4/AlwaysSample-10                   9.00 ± 0%      9.00 ± 0%    ~     (all equal)
SpanWithEvents_4/NeverSample-10                    2.00 ± 0%      2.00 ± 0%    ~     (all equal)
SpanWithEvents_8/AlwaysSample-10                   14.0 ± 0%      14.0 ± 0%    ~     (all equal)
SpanWithEvents_8/NeverSample-10                    2.00 ± 0%      2.00 ± 0%    ~     (all equal)
SpanWithEvents_WithStackTrace/AlwaysSample-10      5.00 ± 0%      5.00 ± 0%    ~     (all equal)
SpanWithEvents_WithStackTrace/NeverSample-10       3.00 ± 0%      3.00 ± 0%    ~     (all equal)
SpanWithEvents_WithTimestamp/AlwaysSample-10       6.00 ± 0%      6.00 ± 0%    ~     (all equal)
SpanWithEvents_WithTimestamp/NeverSample-10        4.00 ± 0%      4.00 ± 0%    ~     (all equal)
```

maisem commented:
@MadVikingGod what go version are you using?

MadVikingGod commented:
It might have been CPU contention that put it in the 25% range, and I will rerun it.  But even your data shows there is a 1-5% performance regression.  So the question I have is what does this patch bring to accept a regression?

MadVikingGod commented:
After Reruning with similar results, I think my question still stands.  To progress this you either need to rework it so that the pool performs equal or better under normal benchmarks, or there needs some compelling reason to accept this.

```
benchstat bench-main bench-3134 
name                                          old time/op    new time/op    delta
SpanProcessor-4                                 13.2µs ± 6%    13.7µs ± 3%  +3.49%  (p=0.019 n=10+10)
SpanProcessorVerboseLogging-4                   14.1µs ± 9%    14.4µs ± 7%    ~     (p=0.256 n=10+10)
SpanLimits/AttributeValueLengthLimit-4          10.8µs ± 7%    10.8µs ± 2%    ~     (p=0.214 n=9+9)
SpanLimits/AttributeCountLimit-4                9.40µs ± 5%    9.58µs ± 6%    ~     (p=0.280 n=10+10)
SpanLimits/EventCountLimit-4                    10.3µs ± 9%    10.3µs ± 4%    ~     (p=0.743 n=9+8)
SpanLimits/LinkCountLimit-4                     10.4µs ± 8%    10.7µs ± 9%    ~     (p=0.190 n=10+10)
SpanLimits/AttributePerEventCountLimit-4        10.7µs ± 8%    10.9µs ± 4%    ~     (p=0.113 n=10+9)
SpanLimits/AttributePerLinkCountLimit-4         10.9µs ± 7%    10.8µs ± 4%    ~     (p=0.661 n=9+10)
SpanSetAttributesOverCapacity-4                 2.25µs ± 6%    2.34µs ± 4%  +3.79%  (p=0.003 n=10+10)
StartEndSpan/AlwaysSample-4                      691ns ± 7%     711ns ± 3%    ~     (p=0.156 n=10+9)
StartEndSpan/NeverSample-4                       320ns ± 4%     346ns ± 6%  +8.30%  (p=0.000 n=10+10)
SpanWithAttributes_4/AlwaysSample-4             1.42µs ± 7%    1.40µs ± 2%    ~     (p=0.400 n=10+9)
SpanWithAttributes_4/NeverSample-4               505ns ± 6%     537ns ± 9%  +6.51%  (p=0.004 n=10+10)
SpanWithAttributes_8/AlwaysSample-4             1.96µs ±10%    1.97µs ± 2%    ~     (p=0.169 n=9+9)
SpanWithAttributes_8/NeverSample-4               672ns ± 6%     704ns ± 5%  +4.80%  (p=0.003 n=10+10)
SpanWithAttributes_all/AlwaysSample-4           1.80µs ± 8%    1.80µs ± 9%    ~     (p=0.987 n=10+10)
SpanWithAttributes_all/NeverSample-4             540ns ± 6%     563ns ± 5%  +4.34%  (p=0.004 n=9+10)
SpanWithAttributes_all_2x/AlwaysSample-4        2.70µs ± 5%    2.78µs ± 7%    ~     (p=0.139 n=9+10)
SpanWithAttributes_all_2x/NeverSample-4          761ns ± 4%     794ns ± 5%  +4.23%  (p=0.001 n=9+10)
SpanWithEvents_4/AlwaysSample-4                 1.70µs ± 9%    1.77µs ± 7%  +4.00%  (p=0.035 n=10+10)
SpanWithEvents_4/NeverSample-4                   329ns ± 3%     354ns ± 4%  +7.36%  (p=0.000 n=9+10)
SpanWithEvents_8/AlwaysSample-4                 2.59µs ± 7%    2.68µs ± 2%  +3.52%  (p=0.010 n=10+9)
SpanWithEvents_8/NeverSample-4                   342ns ± 4%     357ns ± 5%  +4.54%  (p=0.002 n=10+10)
SpanWithEvents_WithStackTrace/AlwaysSample-4    1.03µs ±10%    1.06µs ± 4%    ~     (p=0.050 n=10+10)
SpanWithEvents_WithStackTrace/NeverSample-4      369ns ± 4%     398ns ± 6%  +7.86%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/AlwaysSample-4     1.02µs ± 5%    1.05µs ± 5%  +2.53%  (p=0.040 n=9+9)
SpanWithEvents_WithTimestamp/NeverSample-4       418ns ± 7%     441ns ± 3%  +5.48%  (p=0.002 n=10+9)
TraceID_DotString-4                             88.7ns ± 2%    90.4ns ± 5%    ~     (p=0.065 n=9+10)
SpanID_DotString-4                              62.8ns ± 3%    64.6ns ± 2%  +2.81%  (p=0.003 n=9+9)

name                                          old alloc/op   new alloc/op   delta
SpanProcessor-4                                 8.75kB ± 0%    8.77kB ± 0%  +0.21%  (p=0.000 n=10+10)
SpanProcessorVerboseLogging-4                   9.91kB ± 0%    9.93kB ± 0%  +0.13%  (p=0.000 n=10+8)
SpanLimits/AttributeValueLengthLimit-4          10.6kB ± 0%    10.6kB ± 0%  +0.21%  (p=0.000 n=10+10)
SpanLimits/AttributeCountLimit-4                8.58kB ± 0%    8.59kB ± 0%  +0.20%  (p=0.000 n=10+10)
SpanLimits/EventCountLimit-4                    10.3kB ± 0%    10.3kB ± 0%  +0.20%  (p=0.000 n=10+10)
SpanLimits/LinkCountLimit-4                     10.3kB ± 0%    10.3kB ± 0%  +0.20%  (p=0.000 n=10+10)
SpanLimits/AttributePerEventCountLimit-4        10.5kB ± 0%    10.5kB ± 0%  +0.20%  (p=0.000 n=10+10)
SpanLimits/AttributePerLinkCountLimit-4         10.5kB ± 0%    10.5kB ± 0%  +0.20%  (p=0.000 n=10+10)
SpanSetAttributesOverCapacity-4                   496B ± 0%      496B ± 0%    ~     (all equal)
StartEndSpan/AlwaysSample-4                       432B ± 0%      432B ± 0%    ~     (all equal)
StartEndSpan/NeverSample-4                        128B ± 0%      128B ± 0%    ~     (all equal)
SpanWithAttributes_4/AlwaysSample-4             1.14kB ± 0%    1.14kB ± 0%  +0.18%  (p=0.000 n=10+8)
SpanWithAttributes_4/NeverSample-4                384B ± 0%      384B ± 0%    ~     (all equal)
SpanWithAttributes_8/AlwaysSample-4             1.90kB ± 0%    1.91kB ± 0%  +0.16%  (p=0.000 n=10+10)
SpanWithAttributes_8/NeverSample-4                640B ± 0%      641B ± 0%  +0.16%  (p=0.000 n=10+9)
SpanWithAttributes_all/AlwaysSample-4           1.71kB ± 0%    1.72kB ± 0%  +0.18%  (p=0.000 n=10+8)
SpanWithAttributes_all/NeverSample-4              448B ± 0%      448B ± 0%    ~     (all equal)
SpanWithAttributes_all_2x/AlwaysSample-4        3.06kB ± 0%    3.06kB ± 0%  +0.17%  (p=0.000 n=10+10)
SpanWithAttributes_all_2x/NeverSample-4           768B ± 0%      769B ± 0%  +0.13%  (p=0.000 n=10+10)
SpanWithEvents_4/AlwaysSample-4                   864B ± 0%      865B ± 0%  +0.12%  (p=0.000 n=10+10)
SpanWithEvents_4/NeverSample-4                    128B ± 0%      128B ± 0%    ~     (all equal)
SpanWithEvents_8/AlwaysSample-4                 1.31kB ± 0%    1.31kB ± 0%  +0.15%  (p=0.000 n=10+10)
SpanWithEvents_8/NeverSample-4                    128B ± 0%      128B ± 0%    ~     (all equal)
SpanWithEvents_WithStackTrace/AlwaysSample-4      544B ± 0%      545B ± 0%  +0.11%  (p=0.011 n=10+10)
SpanWithEvents_WithStackTrace/NeverSample-4       144B ± 0%      144B ± 0%    ~     (all equal)
SpanWithEvents_WithTimestamp/AlwaysSample-4       568B ± 0%      569B ± 0%  +0.18%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/NeverSample-4        168B ± 0%      168B ± 0%    ~     (all equal)

<<< Rmoved allocs/op because they were all equal >>>
```

maisem commented:
@MadVikingGod it definitely does look slightly worse for sequential generation, but it makes a massive difference in concurrent generation.

For more context, Tailscale uses its own fork of Go and in that we have added a few metrics around mutex contention (https://github.com/tailscale/go/commit/effe2d16a9ed6c4fd97361e8090bb22acc221075). We noticed that right after we added tracing the mutex contention skyrocketed and kept getting worse as we continued to add more spans.

<img width="1849" alt="image" src="https://user-images.githubusercontent.com/3953239/188017636-cb0462b6-c218-4690-bee0-76fe5a689ff4.png">

Further investigation with some Go's mutex profiling showed this one mutex as being the culprit

<img width="615" alt="image" src="https://user-images.githubusercontent.com/3953239/188017133-eb1ec054-25f4-4285-8922-954bd2640a94.png">

When we deployed this fix last night to prod from our fork (tailscale/opentelemetry-go#3) we immediately saw the contention drop.

![image](https://user-images.githubusercontent.com/3953239/188017246-159c7b7a-afcf-4d62-9f62-d533a7dce79e.png)

hanyuancheung reviewed (approved):
Great work! lgtm.

MadVikingGod commented:
I'm going to try and restate the evidence you've provided.  You have instrumented your mutexes, and then made a change to reduce the number of mutexes, and discovered that you don't spend as much time in mutexes.  I don't disagree with any of that.

But what you failed to show is that by trading a mutex for additional allocations you have improved the performance of creating spans.  When probed on it, your data, and mine, show that this is worse in the base case, 1 thread, and you have week data in your use case, lots of threads.

Things that might help for moving this forward:
1. Explore ways to make the base case better.  I would suggest preallocating one or more RNGs into the pool.  This could even turn into an optimization you might want to make if you make the size of the preallocation a const that you override with a build flag.
2. Build a benchmark demonstrating that this change improves span creation performance in a wide range of cases.  If it's worse in 1 thread, the same in 3, and better at 5 it's a lot easier to take a change than if you need 100 threads to see the effect.

maisem commented:
> Explore ways to make the base case better. I would suggest preallocating one or more RNGs into the pool. This could even turn into an optimization you might want to make if you make the size of the preallocation a const that you override with a build flag.

Preloading the pool has no impact. There is a potential simple improvement that can be made using `mu.TryLock` however that was added in Go 1.18 and this project has a minimal version requirement of Go 1.17. The complexity of the other approaches is not worth it IMO.

> Build a benchmark demonstrating that this change improves span creation performance in a wide range of cases. If it's worse in 1 thread, the same in 3, and better at 5 it's a lot easier to take a change than if you need 100 threads to see the effect.

Added more sub-benchmarks to demonstrate improvement across multiple concurrent worker counts.
There is admittedly a degradation when there is only one worker, but arguably in that case you are not running in a performance critical system to begin with.

For all other cases the improvement is substantial.

```
name                      old time/op  new time/op  delta
SpanIDGeneration/1-10     35.0ns ± 0%  41.2ns ± 0%  +17.83%  (p=0.000 n=9+10)
SpanIDGeneration/2-10     91.6ns ± 1%  46.3ns ± 3%  -49.40%  (p=0.000 n=9+10)
SpanIDGeneration/3-10      120ns ± 1%    51ns ±10%  -57.37%  (p=0.000 n=9+10)
SpanIDGeneration/5-10      134ns ± 0%    66ns ± 0%  -50.57%  (p=0.000 n=9+9)
SpanIDGeneration/10-10     157ns ± 1%    64ns ± 1%  -59.51%  (p=0.000 n=10+10)
SpanIDGeneration/100-10    170ns ± 1%    64ns ± 1%  -62.49%  (p=0.000 n=10+10)
SpanIDGeneration/1000-10   172ns ± 1%    64ns ± 1%  -63.08%  (p=0.000 n=10+10)
```

MrAlias commented:
> this project has a minimal version requirement of Go 1.17

FWIW, we plan to drop 1.17 support (and have 1.18 as minimum) after the [next release](https://github.com/open-telemetry/opentelemetry-go/milestone/26).

maisem commented:
> here is a potential simple improvement that can be made using `mu.TryLock` however that was added in Go 1.18 and this project has a minimal version requirement of Go 1.17.

That is better than the current implementation
```
name                      old time/op  new time/op  delta
SpanIDGeneration/1-10     35.0ns ± 0%  34.2ns ± 0%   -2.36%  (p=0.000 n=9+9)
SpanIDGeneration/2-10     91.6ns ± 1%  54.9ns ± 2%  -40.08%  (p=0.000 n=9+9)
SpanIDGeneration/3-10      120ns ± 1%    45ns ± 1%  -62.68%  (p=0.000 n=9+9)
SpanIDGeneration/5-10      134ns ± 0%    75ns ± 2%  -44.10%  (p=0.000 n=9+10)
SpanIDGeneration/10-10     157ns ± 1%    77ns ± 2%  -50.87%  (p=0.000 n=10+10)
SpanIDGeneration/100-10    170ns ± 1%    79ns ± 1%  -53.36%  (p=0.000 n=10+10)
SpanIDGeneration/1000-10   172ns ± 1%    81ns ± 0%  -53.19%  (p=0.000 n=10+8)
```
But worse than what this PR does

```
name                      old time/op  new time/op  delta
SpanIDGeneration/1-10     41.2ns ± 0%  34.2ns ± 0%  -17.13%  (p=0.000 n=10+9)
SpanIDGeneration/2-10     46.3ns ± 3%  54.9ns ± 2%  +18.43%  (p=0.000 n=10+9)
SpanIDGeneration/3-10     51.2ns ±10%  44.8ns ± 1%  -12.45%  (p=0.000 n=10+9)
SpanIDGeneration/5-10     66.2ns ± 0%  74.8ns ± 2%  +13.08%  (p=0.000 n=9+10)
SpanIDGeneration/10-10    63.7ns ± 1%  77.2ns ± 2%  +21.34%  (p=0.000 n=10+10)
SpanIDGeneration/100-10   63.7ns ± 1%  79.2ns ± 1%  +24.32%  (p=0.000 n=10+10)
SpanIDGeneration/1000-10  63.7ns ± 1%  80.7ns ± 0%  +26.78%  (p=0.000 n=10+8)
```

maisem commented:
Also there are no allocation changes
```
goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/otel/sdk/trace
BenchmarkSpanIDGeneration/1-10          30392380                39.44 ns/op            0 B/op          0 allocs/op
BenchmarkSpanIDGeneration/2-10          30977733                33.85 ns/op            0 B/op          0 allocs/op
BenchmarkSpanIDGeneration/3-10          36016962                33.42 ns/op            0 B/op          0 allocs/op
BenchmarkSpanIDGeneration/5-10          17810319                71.84 ns/op            0 B/op          0 allocs/op
BenchmarkSpanIDGeneration/10-10         15673084                83.68 ns/op            0 B/op          0 allocs/op
BenchmarkSpanIDGeneration/100-10                15104115                81.75 ns/op            0 B/op          0 allocs/op
BenchmarkSpanIDGeneration/1000-10               15314407                83.78 ns/op            0 B/op          0 allocs/op
PASS
ok      go.opentelemetry.io/otel/sdk/trace      10.882s
```

maisem commented:
@MadVikingGod ping :)

MadVikingGod commented:
A number of thoughts where brought up in the SIG meeting.

1. This is a PR to change the default.  You are going to need to show that this is not a regression for the current cases, eg. the rest of the benchmarks in the sdk. And that this is a performance improvement in more than just spanID generation, for example that creating spans in parallel is faster.
2. You can already override this using [WithIDGenerator](https://pkg.go.dev/go.opentelemetry.io/otel/sdk@v1.9.0/trace#WithIDGenerator).

maisem commented:
> You can already override this using [WithIDGenerator](https://pkg.go.dev/go.opentelemetry.io/otel/sdk@v1.9.0/trace#WithIDGenerator).

Thanks, TIL!

> This is a PR to change the default.

Okay, if you don't think optimizing for high concurrency by default is the right default, we can close this.

dmathieu commented:
> Okay, if you don't think optimizing for high concurrency by default is the right default, we can close this.

I don't think this is what @MadVikingGod said in his previous comment. We discussed this during last week's SIG meeting (you're welcome to join. You can find the details in the [otel calendar](https://github.com/open-telemetry/community#calendar)), and were definitely not closed to making this change.
My feeling is that one thing which would help a lot would be reproducible data on a standard Go app showing this is indeed a production improvement.

maisem commented:
@dmathieu yes I agree that having a production like env to test these type of changes would indeed be great. However, I don't see the benefit of blocking this change until the overall project has better perf benchmarks.

I've already demonstrated that it had a substantial difference in our production environment. The ask to add a benchmark showing the difference was reasonable and I did that, asking for a whole production-like test bed to demonstrate the change is not.

In regards to the regression in performance in single threaded benchmarks, it is mostly irrelevant as I can't think of any performance critical environment where a 5% serial degradation has any substantial impact AND is not immediately overshadowed by the improvement in concurrent flows. I'd be super curious if such an environment indeed does exist, but even if it does; optimizing for that in a default implementation seems incorrect.

FWIW, I also don't think that result is entirely accurate most likely due to the benchmarks not being configured correctly to not account for setup costs but that's  only speculation at this point.
